### PR TITLE
feat(fsdp_tp): DCP checkpointing + auto-resume + restart-time (validated on Sunspot)

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Adds **sharded checkpointing** to `ezpz.examples.fsdp_tp` so training can save
and resume model/optimizer state — then **validates it end-to-end on real
Sunspot XPU hardware** and documents a restart-under-failure investigation.
Composes with `ezpz launch --auto-retry` (#195): auto-resume-on-startup means
a relaunch just resumes.

Before this, fsdp_tp had **no** checkpointing — every restart began at step 0.

## Feature

**`src/ezpz/examples/_checkpoint.py` (new)** — FSDP2/DTensor-aware DCP:
- `save_checkpoint` — `get_state_dict` + `dcp.save` (each rank writes its
  shard in parallel); rank 0 writes `meta.json`, then a `.complete` marker
  **last** (crash-safety: a save interrupted mid-write is skipped).
- `load_checkpoint` — loads the latest **complete** checkpoint; returns meta
  or `None` (fresh run).
- `latest_checkpoint` — newest `step-<N>` with a `.complete` marker.

**`fsdp_tp.py` wiring** — flags `--ckpt-dir` / `--save-interval` /
`--train-iters` / `--no-resume`; **auto-resume** after the sharded
model+optimizer are built (seeds step/tokens/epoch/offset); loop resumes at
`start_epoch`, skips consumed batches, honors `--train-iters`.

**Restart-time measurement** — a `RESUMED from step=N` log line + a
`train/restart_seconds` metric timed from `main()` entry (so it covers the
full cold path incl. `setup_torch`, not just `dcp.load`).

## Validation — real Sunspot measurements (2 nodes / 24 XPU ranks)

The DCP path was proven on hardware (not just local CPU unit tests):

- **DCP save/resume on XPU** (job `ckpt_validate`): sharded save across 24
  ranks (25 shard files + `.complete` per checkpoint), clean auto-resume from
  step 30, `restart_seconds=10.7s`.
- **Checkpoint-restart experiment** (job 12471687): baseline **5.81 min** vs
  restart **8.04 min** for 3000 steps across **4 real SIGKILL failures**
  (+38% overhead). Per failure: restart **10.4–11.1 s**, lost **57–71 steps**
  (bounded by the 100-step interval).

Artifacts: `experiments/checkpoint-restart/` (driver + reusable plotter +
README with the plot) and a docs guide page,
`docs/guides/checkpoint-restart.md` ("💾 Checkpoint & Restart Under Failure").

## Tests
`test_checkpoint.py` (7): `latest_checkpoint` ordering + incomplete-skip;
stale-marker-cleared-on-overwrite; full save→load round-trip (params + AdamW
step state + meta); process-group teardown so it doesn't pollute other
modules. **ruff clean; affected suites pass.**

## Review
All review findings addressed (Codex + Copilot + Sourcery): restart timer now
covers distributed setup; stale `.complete` cleared before overwrite; per-rank
log spam gated to rank 0; exit-code robustness; test env/PG isolation. 0
unresolved threads; CI green.

## Scope / non-goals
Sharded checkpoints only (no gathered save); synchronous `dcp.save` (async is
a follow-up); same-parallelism resume (no reshard). The reference
fault-tolerance chart's TorchPass/TorchFT lines are separate frameworks not
integrated into ezpz and are out of scope.

## Test plan
- [x] Local: pytest (checkpoint round-trip, isolation) + ruff clean
- [x] Sunspot: sharded DCP save/resume on XPU; 4-failure restart experiment with real restart-seconds + lost-steps
- [x] Docs guide + reusable experiment artifact

### CODE REVIEW

File: src/ezpz/examples/fsdp_tp.py [hunk 1]
```diff
@@ -2155,6 +2201,11 @@ def train(
             When non-None, ``profiler.step()`` is called once per training
             step so the schedule (wait/warmup/active/repeat) advances.
     """
+    # Training-entry timestamp for the restart-time metric. Captured first so
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Reasonable — the loop grew a few resume/stop/save locals. Kept the wiring inline for this PR (it stays close to the loop it drives) but pulled shared rank/complete-marker logic into _checkpoint.py helpers. Will extract a ResumeState dataclass as a follow-up if the loop grows further.

File: src/ezpz/examples/_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    import torch.distributed as dist
+    import torch.distributed.checkpoint as dcp
+    from torch.distributed.checkpoint.state_dict import get_state_dict
+
+    out = _step_dir(ckpt_dir, step)
+    is_rank0 = (not dist.is_available()) or (not dist.is_initialized()) or (
+        dist.get_rank() == 0
+    )
+    if is_rank0:
+        out.mkdir(parents=True, exist_ok=True)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/_checkpoint.py:101]: Fixed in efa465f — save_checkpoint removes any existing .complete marker BEFORE dcp.save, so a crash mid-overwrite leaves the dir markerless and latest_checkpoint() falls back to the previous good step. Added a test that raises mid-dcp.save and asserts latest_checkpoint() returns None.

File: src/ezpz/examples/fsdp_tp.py [hunk 3]
```diff
@@ -2155,6 +2201,11 @@ def train(
...
+    # train/restart_seconds includes model build + dcp.load, i.e. the full cold
+    # path an --auto-retry relaunch pays before its first productive step.
+    _train_t0 = perf_counter()
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/fsdp_tp.py]: Good catch, and important for the restart-time experiment. Fixed in efa465f: main() captures t0 BEFORE setup_torch(), threaded into train(process_start=t0), so train/restart_seconds now includes distributed init (the dominant cold-start cost). Falls back to now() when train() is called directly.

File: tests/test_checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    if dist.is_available() and not dist.is_initialized():
+        import os
+
+        os.environ.setdefault("MASTER_ADDR", "127.0.0.1")
+        os.environ.setdefault("MASTER_PORT", "29555")
+        os.environ.setdefault("RANK", "0")
+        os.environ.setdefault("WORLD_SIZE", "1")
+        dist.init_process_group(backend="gloo", rank=0, world_size=1)
+    return dist
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_checkpoint.py:70]: Resolved in 9a2e5aa (the test-pollution fix): _init_single_rank_pg no longer touches os.environ — it passes rank/world_size + a tcp init_method directly to init_process_group, so no MASTER_PORT setdefault to inherit/collide.

File: tests/test_checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        model, opt = _Tiny.build(torch)
+        # Take a couple of steps so optimizer state (exp_avg, etc.) is non-empty.
+        for _ in range(3):
+            opt.zero_grad()
+            out = model(torch.randn(4, 8))
+            out.sum().backward()
+            opt.step()
+
+        ref_params = [p.detach().clone() for p in model.parameters()]
+        ref_step0 = opt.state[next(iter(opt.state))]["step"].clone()
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_checkpoint.py]: Fixed in efa465f — the test normalizes AdamW per-param step to int via int(s.item() if hasattr(s,"item") else s) instead of .clone()/torch.equal, robust to int-or-0dim-tensor.

File: src/ezpz/examples/_checkpoint.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    if is_rank0:
+        payload = dict(meta or {})
+        payload["step"] = step
+        (out / _META_FILE).write_text(json.dumps(payload, indent=2))
+        # Marker LAST: its presence is the "this checkpoint is complete and
+        # safe to resume from" contract that latest_checkpoint() relies on.
+        (out / _COMPLETE_MARKER).write_text("")
+    _barrier()
+    logger.info("saved checkpoint: %s", out)
+    return out
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/_checkpoint.py:129]: Fixed in efa465f — the "saved checkpoint" line is gated behind _is_rank0().

File: src/ezpz/examples/_checkpoint.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    meta_path = latest / _META_FILE
+    meta: dict[str, Any] = {}
+    if meta_path.exists():
+        try:
+            meta = json.loads(meta_path.read_text())
+        except (ValueError, OSError):
+            logger.warning("could not parse %s; resuming with step only", meta_path)
+    meta.setdefault("step", int(latest.name[len(_STEP_PREFIX):]))
+    logger.info("resumed from checkpoint: %s (step=%s)", latest, meta.get("step"))
+    return meta
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/examples/_checkpoint.py:182]: Fixed in efa465f — the meta-parse warning and "resumed from checkpoint" info are gated behind _is_rank0().
